software/contrib/lutra.py

```diff
@@ -0,0 +1,327 @@
+            theta = (self.current_tick / self.cycle_ticks) * 2 * math.pi
+            volts = (-math.cos(theta) + 1) / 2 * MAX_OUTPUT_VOLTAGE
+        elif self.shape == self.WAVE_SHAPE_SQUARE:
+            if self.current_tick < (self.cycle_ticks >> 1):
+                volts = MAX_OUTPUT_VOLTAGE
+            else:
+                volts = 0
+        elif self.shape == self.WAVE_SHAPE_TRIANGLE:
+            half_cycle_ticks = (self.cycle_ticks >> 1)
+            if self.current_tick < half_cycle_ticks:
+                volts = self.current_tick / half_cycle_ticks * MAX_OUTPUT_VOLTAGE
+            else:
+                volts = MAX_OUTPUT_VOLTAGE - (self.current_tick - half_cycle_ticks) / half_cycle_ticks * MAX_OUTPUT_VOLTAGE
+        elif self.shape == self.WAVE_SHAPE_SAW:
+            volts = MAX_OUTPUT_VOLTAGE - self.current_tick / self.cycle_ticks * MAX_OUTPUT_VOLTAGE
+        elif self.shape == self.WAVE_SHAPE_RAMP:
+            volts = self.current_tick / self.cycle_ticks * MAX_OUTPUT_VOLTAGE
+        else:
+            volts = 0
+
+        self.current_tick = self.current_tick + 1
+        if self.current_tick >= self.cycle_ticks:
+            self.current_tick = 0
+
+        self.cv_output.voltage(volts)
+        return volts
+
+class Lutra(EuroPiScript):
+    """The main class for this script; handles running the main loop, configuring I/O, loading, and saving state.
+    """
+
+    # We support CV control over either LFO speed OR LFO spread via AIN.  This option is not exposed through a menu
+    # and must be configured via the config file.  See lutra.md for details
+    AIN_MODE_SPREAD = 0
+    AIN_MODE_SPEED = 1
+    AIN_MODE_NAMES = [
+        "Spread",
+        "Speed"
+    ]
+
+    # The maximum and minimum cycle time for the LFOs
+    MIN_CYCLE_TICKS = 250
+    MAX_CYCLE_TICKS = 10000
+
+    # Maximum wave speed multipliers relative to cv1
+    MAX_SPEED_MULTIPLIERS = [
+        1/1,
+        6/5,
+        5/4,
+        4/3,
+        3/2,
+        2/1
+    ]
+
+    def __init__(self):
+        """Constructor
+        """
+        super().__init__()
+
+        self.waves = [
+            WaveGenerator(cv) for cv in cvs
+        ]
+        self.config_dirty = False
+        self.hold_low = False
+        self.last_wave_change_at = time.ticks_ms()
+
+        # Connect the B2 handler to our digital input helper
+        # B1 and DIN are handled differently since they interact with each other
+        # See @wave_generator_thread
+        self.digital_input_state = DigitalInputHelper(
+            on_b2_rising = self.on_b2_rising
+        )
+
+        # Save the last screen width's worth of output voltages converted to pixel heights
+        # This speeds up rendering
+        self.display_pixels = [
+            [] for cv in cvs
+        ]
+
+        self.pixel_lock = _thread.allocate_lock()
+
+        self.load()
+
+    def load(self):
+        """Load and apply the saved configuration
```

Comment:
[required] The configuration file and the saved state file are two different things in the europi firmware. It's confusing to have the documentation here call it a configuration file and then read from the saved state file. Ditto for the `save()` method below.

Reply:
Like I said above, the saved_state file (in my interpretation) _is_ the same thing as a per-script configuration file.

Reply:
Fixed